### PR TITLE
Resolve users when multiple user stores are configured in the authenticator.

### PR DESCRIPTION
**Pull request by Yasasr1**
### Purpose
This improves the authenticator to resolve the user correctly with multi attribute login when multiple user stores are configured in the authenticator.

### CODE REVIEW

File: components/org.wso2.carbon.identity.application.authenticator.iwa/src/main/java/org/wso2/carbon/identity/application/authenticator/iwa/IWAFederatedAuthenticator.java [hunk 1]
```diff
@@ -212,31 +212,46 @@ private IWAAuthenticatedUserBean userInformationInListedUserStores(String authen
...
-                    break;
+        authenticatedUserBean.setTenantDomain(tenantDomain);
+        authenticatedUserBean.setUser(authenticatedUserName);
+        MultiAttributeLoginService multiAttributeLoginService =
+                IWAServiceDataHolder.getInstance().getMultiAttributeLoginService();
+
+        if (multiAttributeLoginService.isEnabled(tenantDomain)) {
+            ResolvedUserResult resolvedUserResult = multiAttributeLoginService.resolveUser(
+                    authenticatedUserName, tenantDomain);
+            if (resolvedUserResult != null &&
```
Comment (indeewari) [inline components/org.wso2.carbon.identity.application.authenticator.iwa/src/main/java/org/wso2/carbon/identity/application/authenticator/iwa/IWAFederatedAuthenticator.java:228]: If there is no user found, authenticatedUserBean will be returned as userExists = false ? and the later logic will handle there is no result found via userExists variable ?

File: components/org.wso2.carbon.identity.application.authenticator.iwa/src/main/java/org/wso2/carbon/identity/application/authenticator/iwa/IWAFederatedAuthenticator.java [hunk 2]
```diff
@@ -212,31 +212,46 @@ private IWAAuthenticatedUserBean userInformationInListedUserStores(String authen
...
+                    ResolvedUserResult.UserResolvedStatus.SUCCESS.equals(resolvedUserResult.getResolvedStatus()) &&
+                    userStoreDomains.toUpperCase().contains(resolvedUserResult.getUser().getUserStoreDomain())) {
```
Comment (indeewari) [inline components/org.wso2.carbon.identity.application.authenticator.iwa/src/main/java/org/wso2/carbon/identity/application/authenticator/iwa/IWAFederatedAuthenticator.java]: A "contains" check on a comma-separated string may also match substrings.
i.e DEV,INTENALDEV,EXTERNALDEV.

File: components/org.wso2.carbon.identity.application.authenticator.iwa/src/main/java/org/wso2/carbon/identity/application/authenticator/iwa/IWAFederatedAuthenticator.java [hunk 3]
```diff
@@ -212,31 +212,46 @@ private IWAAuthenticatedUserBean userInformationInListedUserStores(String authen
...
+            String userStoreDomainForUser = null;
+            try {
+                for (String userStoreDomain : userStoreDomains.split(",")) {
+                    if (isUserExistsInUserStore(authenticatedUserName, tenantDomain, userStoreDomain.trim())) {
+                        isUserExists = true;
+                        userStoreDomainForUser = userStoreDomain.trim();
+                        break;
+                    }
+                }
+                authenticatedUserBean.setUserExists(isUserExists);
```
Comment (indeewari) [inline components/org.wso2.carbon.identity.application.authenticator.iwa/src/main/java/org/wso2/carbon/identity/application/authenticator/iwa/IWAFederatedAuthenticator.java]: Can't we relocate these two lines

authenticatedUserBean.setUserExists(isUserExists);
authenticatedUserBean.setUserStoreDomain(userStoreDomainForUser);

into the if part ? i.e

for (String userStoreDomain : userStoreDomains.split(",")) {
                    if (isUserExistsInUserStore(authenticatedUserName, tenantDomain, userStoreDomain.trim())) {
authenticatedUserBean.setUserExists(true);
authenticatedUserBean.setUserStoreDomain(userStoreDomain.trim());
                        break;
                    }
                }

File: components/org.wso2.carbon.identity.application.authenticator.iwa/src/main/java/org/wso2/carbon/identity/application/authenticator/iwa/IWAFederatedAuthenticator.java [hunk 4]
```diff
@@ -212,31 +212,46 @@ private IWAAuthenticatedUserBean userInformationInListedUserStores(String authen
...
+                authenticatedUserBean.setUserStoreDomain(userStoreDomainForUser);
 
-            authenticatedUserBean.setTenantDomain(tenantDomain);
-            if (StringUtils.isEmpty(authenticatedUserBean.getUser())) {
-                authenticatedUserBean.setUser(authenticatedUserName);
+            } catch (AuthenticationFailedException e) {
+                String msg = "IWAApplicationAuthenticator failed to find the user:%s of tenantDomain=%s in neither " +
```
Comment (indeewari) [inline components/org.wso2.carbon.identity.application.authenticator.iwa/src/main/java/org/wso2/carbon/identity/application/authenticator/iwa/IWAFederatedAuthenticator.java:247]: Don't we have error constant class to maintain these error messages and proper codes ?

File: components/org.wso2.carbon.identity.application.authenticator.iwa/src/main/java/org/wso2/carbon/identity/application/authenticator/iwa/IWAFederatedAuthenticator.java (hunk 1 above)
Comment (Yasasr1) [reply components/org.wso2.carbon.identity.application.authenticator.iwa/src/main/java/org/wso2/carbon/identity/application/authenticator/iwa/IWAFederatedAuthenticator.java:228]: Yes, this is handled here https://github.com/wso2-extensions/identity-local-auth-iwa-kerberos/blob/48bf34fd07cb6748cba41a8efa42d856aee38c82/components/org.wso2.carbon.identity.application.authenticator.iwa/src/main/java/org/wso2/carbon/identity/application/authenticator/iwa/IWAFederatedAuthenticator.java#L123

File: components/org.wso2.carbon.identity.application.authenticator.iwa/src/main/java/org/wso2/carbon/identity/application/authenticator/iwa/IWAFederatedAuthenticator.java (hunk 4 above)
Comment (Yasasr1) [reply components/org.wso2.carbon.identity.application.authenticator.iwa/src/main/java/org/wso2/carbon/identity/application/authenticator/iwa/IWAFederatedAuthenticator.java:247]: Error messages are not maintained in a constant class in this component

Comment (indeewari) [reply components/org.wso2.carbon.identity.application.authenticator.iwa/src/main/java/org/wso2/carbon/identity/application/authenticator/iwa/IWAFederatedAuthenticator.java:247]: Shall we add an error class with descriptions and codes ?

File: components/org.wso2.carbon.identity.application.authenticator.iwa/src/main/java/org/wso2/carbon/identity/application/authenticator/iwa/IWAFederatedAuthenticator.java (hunk 2 above)
Comment (Yasasr1) [reply components/org.wso2.carbon.identity.application.authenticator.iwa/src/main/java/org/wso2/carbon/identity/application/authenticator/iwa/IWAFederatedAuthenticator.java]: updated with [6f93b37](https://github.com/wso2-extensions/identity-local-auth-iwa-kerberos/pull/77/commits/6f93b37a06ebaf7283992851d1692a50f00853bd)

File: components/org.wso2.carbon.identity.application.authenticator.iwa/src/main/java/org/wso2/carbon/identity/application/authenticator/iwa/IWAFederatedAuthenticator.java (hunk 3 above)
Comment (Yasasr1) [reply components/org.wso2.carbon.identity.application.authenticator.iwa/src/main/java/org/wso2/carbon/identity/application/authenticator/iwa/IWAFederatedAuthenticator.java]: updated with [6f93b37](https://github.com/wso2-extensions/identity-local-auth-iwa-kerberos/pull/77/commits/6f93b37a06ebaf7283992851d1692a50f00853bd)

File: components/org.wso2.carbon.identity.application.authenticator.iwa/src/main/java/org/wso2/carbon/identity/application/authenticator/iwa/IWAFederatedAuthenticator.java [hunk 5]
```diff
@@ -212,31 +214,43 @@ private IWAAuthenticatedUserBean userInformationInListedUserStores(String authen
...
-                if (authenticatedUserBean.isUserExists()) {
-                    userStoreDomainForUser = userStoreDomain.trim();
-                    break;
+        authenticatedUserBean.setTenantDomain(tenantDomain);
+        authenticatedUserBean.setUser(authenticatedUserName);
+        MultiAttributeLoginService multiAttributeLoginService =
+                IWAServiceDataHolder.getInstance().getMultiAttributeLoginService();
+
+        if (multiAttributeLoginService.isEnabled(tenantDomain)) {
+            List<String> userStoreDomainsList = Arrays.stream(userStoreDomains.toUpperCase().split(","))
```
Comment (indeewari) [inline components/org.wso2.carbon.identity.application.authenticator.iwa/src/main/java/org/wso2/carbon/identity/application/authenticator/iwa/IWAFederatedAuthenticator.java:224]: Is userStoreDomains null safe ?

Comment (Yasasr1) [reply components/org.wso2.carbon.identity.application.authenticator.iwa/src/main/java/org/wso2/carbon/identity/application/authenticator/iwa/IWAFederatedAuthenticator.java:224]: yes, https://github.com/wso2-extensions/identity-local-auth-iwa-kerberos/blob/6f93b37a06ebaf7283992851d1692a50f00853bd/components/org.wso2.carbon.identity.application.authenticator.iwa/src/main/java/org/wso2/carbon/identity/application/authenticator/iwa/IWAFederatedAuthenticator.java#L115

**Review by indeewari — APPROVED** [review]

**Pull request merged**